File: backend/app/schemas/ocr.py
"""OCR 相关数据模式."""

from pydantic import BaseModel, Field


class OCRProcessRequest(BaseModel):
    """OCR 处理请求（手动输入模式）."""

    text: str = Field(..., min_length=10, description="用户粘贴的试卷文本")
    source: str | None = Field(default=None, description="来源（如：2024年中考真题）")
    source_year: int | None = None
    region: str | None = None
    exam_type: str | None = None
    auto_tag: bool = Field(default=True, description="是否自动标注知识点和难度")


class ExtractedQuestionData(BaseModel):
    """提取的题目数据."""

    number: str
    content_latex: str
    question_type: str = "short_answer"
    options: list[dict] | None = None
    answer_latex: str | None = None
    knowledge_points: list[str] = []
    estimated_difficulty: float = 3.0


class OCRResultResponse(BaseModel):
    """OCR 处理结果."""

    status: str
    raw_text: str = ""
    cleaned_text: str = ""
    questions: list[ExtractedQuestionData] = []
    corrections: list[dict] = []
    confidence: float = 0.0
    notes: str = ""
    error: str | None = None
    saved_question_ids: list[int] = []

File: frontend/src/api/ocr.ts
/** OCR 识别 API */

import client from './client';
import type { UnifiedResponse } from '@/types/common';

export interface ExtractedQuestionData {
  number: string;
  content_latex: string;
  question_type: string;
  options: { label: string; content_latex: string }[] | null;
  answer_latex: string | null;
  knowledge_points: string[];
  estimated_difficulty: number;
}

export interface OCRResultData {
  status: string;
  raw_text: string;
  cleaned_text: string;
  questions: ExtractedQuestionData[];
  corrections: { original: string; corrected: string; reason: string }[];
  confidence: number;
  notes: string;
  error: string | null;
  saved_question_ids: number[];
}

export interface OCRProcessRequest {
  text: string;
  source?: string;
  source_year?: number;
  region?: string;
  exam_type?: string;
  auto_tag?: boolean;
}

/** 手动输入模式：文本提交 */
export async function processText(data: OCRProcessRequest): Promise<OCRResultData> {
  const res = await client.post<UnifiedResponse<OCRResultData>>('/ocr/process', data);
  if (!res.data.data) throw new Error('OCR 处理失败：服务器未返回数据');
  return res.data.data;
}

/** 文件上传模式 */
export async function uploadDocument(
  file: File,
  metadata?: {
    source?: string;
    source_year?: number;
    region?: string;
    exam_type?: string;
    auto_tag?: boolean;
  }
): Promise<OCRResultData> {
  const formData = new FormData();
  formData.append('file', file);
  if (metadata?.source) formData.append('source', metadata.source);
  if (metadata?.source_year) formData.append('source_year', String(metadata.source_year));
  if (metadata?.region) formData.append('region', metadata.region);
  if (metadata?.exam_type) formData.append('exam_type', metadata.exam_type);
  if (metadata?.auto_tag !== undefined) formData.append('auto_tag', String(metadata.auto_tag));

  const res = await client.post<UnifiedResponse<OCRResultData>>('/ocr/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 300000, // 5 分钟超时
  });
  if (!res.data.data) throw new Error('文件上传处理失败：服务器未返回数据');
  return res.data.data;
}
